### DISCUSSION TITLE
How to enable DaisyUI with Nativewind for native

### DISCUSSION THREAD

**Discussion by dohomi** [nativewind/nativewind#223] [Q&A]
Hi Guys,

I successfully enabled DaisUI with the mono repo setup to run equally on Web with Storybook and NextJS. Now I am struggling to see any DaisyUI styles with native (iOS).

I guess this is due to the fact that DaisyUI groups components via class names like `btn btn-primary`. All this names seems to not be part of the generated stylesheet, while TailwindCSS class names like `bg-sky-100` render correctly.

@marklawlor do you have an idea if the class names are ignored with the Nativewind compiler?

**Comment by dohomi** [comment]
I followed the Troubleshooting page and did https://www.nativewind.dev/guides/cli-native#native. Maybe I overlook something obvious but the nativewind-output.js contains all class names while the components still remain unstyled with DaisyUI classNames.
https://github.com/dohomi/nativewind-solito-kitchen-sink

**Reply by dohomi** [reply]
I changed some styles to `!btn` to overcome the issue that the important keyword didn't work for `important:html`. Not sure if that is the key error at the moment

**Reply by dohomi** [reply]
I tracked the issue further down and realised that the nativewind CLI does not include the css-vars which are required for the theming of DaisyUI. 
@marklawlor do you have an idea how to troubleshoot the missing css vars?
https://github.com/dohomi/nativewind-solito-kitchen-sink/blob/master/apps/expo/nativewind-output.js#L8

**Reply by dohomi** [reply]
<img width="776" alt="Screenshot 2022-09-02 at 12 41 53" src="https://user-images.githubusercontent.com/489221/188060471-852c4af2-7059-40d2-a0e8-6d901f77d90e.png">

**Reply by marklawlor** [reply]
So we're probably about a year away from supporting this completely.

CSS vars are a complex problem. Read the caveats section of this to understand.
 https://www.npmjs.com/package/postcss-css-variables#caveats

So we can't preprocess CSS vars, which means we need to process them at runtime (which we need for dynamic cars anyway).

I'd love to get to that point, but these are the milestones between here and there

- [ ] Rewrite CSS parser to be a bespoke parser (it's currently a hack around the library StyledComponents use, which it's great for our use case)
- [ ]  Add support for root level variables (initial support / REM support)
- [ ] Add support for cascading var support (em unit support)
- [ ] Improve media query parsing
- [ ] Proper CSS car support

That's the rough plan, but mixed between are other features people are requesting.

So we'll get there, but not in the short term unfortunately ☹️

**Reply by dohomi** [reply]
@marklawlor thank you so much for your explanations. Are there some "low hanging fruits" possibilities? Something like the Nativewind compiler would blindly keep the var(--c) names while the developer has to add the `:root{}` vars into the root App.jsx by himself? Or am I misunderstanding the concept of how Nativescript compiler works?

**Reply by marklawlor** [reply]
The low hanging fruit is already implemented, it's just not good enough :disappointed: 

https://github.com/marklawlor/nativewind/blob/45b45fe0daba68c608f8bc058ab3c472782513c2/packages/nativewind/src/postcss/index.ts#L1-L23

That tries to resolve CSS variables, dynamic CSS like calc() and make media queries easier to parse. I don't think I can do anything more at the compile stage, hence the need for a more intelligent runtime.

**Reply by dohomi** [reply]
I tried the postcss CLI output and I receive a ton of warnings and basically all css vars are not part of the generated .js file. I will see if I find another way to overwrite all missing css vars.

**Reply by marklawlor** [reply]
I think you may have misunderstood. The reason why there are no CSS vars in the js output, is because they are removed. There's nothing in React Native/NativeWind that understands what a CSS variable is, so they are statically analysed and resolved. You can read the README of the postcss-css-variables that I linked above for more information.

The warnings are probably received were `calc()` functions that were also impossible to statically resolved. Because React Native doesn't have the equivalent of `calc()` they also need to analysed and removed.

Having taken a closer look at DaisyUI, I would need to implement a lot of the CSS spec for it to translate to RN. The goal of this project is to support Tailwind CSS, not the CSS spec, so I'm not sure if I'd ever end up supporting DaisyUI.

**Reply by dohomi** [reply]
Alright thanks for the explanation! I had a misunderstanding through https://blog.logrocket.com/using-css-variables-react-native/ this blog articles that in ReactNative css vars are fully supported. I am new to ReactNative (been using React for years but never natively) so I am still elaborating how to track down the requirements for the upcoming project. Was trying out Dripsy, Tamagui and some other solutions and found Nativewind most appealing due to the big ecosystem of Tailwind. 

I tried in a second branch `flowbite` but I guess they are also using non supported names which breaks Nativewind. I will do more research about this later on.

**Reply by marklawlor** [reply]
That article is very confusing and kinda looks like blog spam. 

Firstly its talking about React Native Web, but it never states that. Then it shows "examples" of RNW being styled, but it never applies the styles to the RNW elements :confused:  It just creates DOM elements and styles them. Or applies styles to the RN elements via global CSS styles - which you can only do on the web.

Plugins like DaisyUI and Flowbite are big plugins that take Tailwind away from being a utility framework and into a component framework. I don't think your going to have much success with these types of plugins as they rely too much on CSS.

**Reply by dohomi** [reply]
Thats exactly what I realised - the blog article uses styled component and doesn't even actually run in ReactNative- lesson learned. The title is very clickbait-ish and I was falling for it :-)

On my first look Flowbite seemed to use solely Tailwind class names but after trying to borrow https://github.com/themesberg/flowbite-react/blob/main/src/lib/theme/default.ts and start using the class names the app stopped building. My guess is, that certain class names are not supported and that was breaking the build system.

**Reply by xWayfinder** [reply]
Hi @marklawlor I stumbled on this 2 year old thread, with a similar idea as @dohomi of using something like daisyUI with nativewind. Just wondering if this is something you and the team are thinking of building out support of and if maybe I could contribute. Thanks again for creating such a great library!


> So we're probably about a year away from supporting this completely.
> 
> CSS vars are a complex problem. Read the caveats section of this to understand. https://www.npmjs.com/package/postcss-css-variables#caveats
> 
> So we can't preprocess CSS vars, which means we need to process them at runtime (which we need for dynamic cars anyway).
> 
> I'd love to get to that point, but these are the milestones between here and there
> 
> * [ ]  Rewrite CSS parser to be a bespoke parser (it's currently a hack around the library StyledComponents use, which it's great for our use case)
> * [ ]   Add support for root level variables (initial support / REM support)
> * [ ]  Add support for cascading var support (em unit support)
> * [ ]  Improve media query parsing
> * [ ]  Proper CSS car support
> 
> That's the rough plan, but mixed between are other features people are requesting.
> 
> So we'll get there, but not in the short term unfortunately ☹️

**Reply by marklawlor** [reply]
CSS Variable support is coming, but DaisyUI uses CSS selectors which I'm not sure we'll ever implement.

There's also just simple things like `btn`. Its "active" styling (the animation when you press in) is `btn:active:focus`. In React Native,`<Pressable />` does not have an `onFocus` event handler so we cannot meet the `:focus` condition (tvOS does have the `onFocus` handler).

DaisyUI also does "HTML" tricks. Like this button isn't really a button, its a `<detail />` or a `<input type="checkbox" />`.

**Reply by xWayfinder** [reply]
Thanks for explaining @marklawlor. It might more sense for us to just imitate aspects of daisyUI with NativeWind and cover the gaps with react native and js.

**Comment by danstepanov** [comment]
Closing this discussion, if this is still relevant, feel free to open a new one.
